### PR TITLE
MAINT: Switch to latest ASV, use mamba solver

### LINKED ISSUE #18849: BUG: `python dev.py bench` fails
Trying to run the benchmarks in WSL seems to create an invalid yaml somewhere along the way, which then fails when trying to create a conda environment based on it.

```
$ python dev.py bench --compare main
[...]
conda env create -f /tmp/tmpop7mwsw6.yml -p /mnt/d/upstreams/scipy/benchmarks/env/dc4107ba04c9faab49c4c6d6f6b22097 --force 
   SpecNotFound: /tmp/tmpop7mwsw6.yml is not a valid yaml file.
```

Not sure if this is a problem in asv, conda, WSL, or our dev.py interface, but it keeps me from running benchmarks. 😑 

<details>
<summary>Stacktrace</summary>

```
$ python dev.py bench --compare main
💻  ninja -C /mnt/d/upstreams/scipy/build -j8
ninja: Entering directory `/mnt/d/upstreams/scipy/build'
[4/4] Generating scipy/generate-version with a custom command
Build OK
💻  meson install -C build --only-changed
Installing, see meson-install.log...
    ... installation in progress (0:01:00.304647 elapsed)
    ... installation in progress (0:02:00.372027 elapsed)
    ... installation in progress (0:03:00.858237 elapsed)
    ... installation in progress (0:04:03.360638 elapsed)
    ... installation in progress (0:05:03.789973 elapsed)
    ... installation in progress (0:06:04.172270 elapsed)
    ... installation in progress (0:07:04.543198 elapsed)
    ... installation in progress (0:08:06.025270 elapsed)
    ... installation in progress (0:09:06.506423 elapsed)
    ... installation in progress (0:10:06.932967 elapsed)
Installation OK
SciPy from development installed path at: /mnt/d/upstreams/scipy/build-install/lib/python3.10/site-packages
********************************************************************************
WARNING: you have uncommitted changes --- these will NOT be benchmarked!
********************************************************************************
· No `environment_type` specified in asv.conf.json. This will be required in the future.
· Creating environments.
·· Error running /home/user/miniforge3/bin/conda env create -f /tmp/tmpop7mwsw6.yml -p /mnt/d/upstreams/scipy/benchmarks/env/dc4107ba04c9faab49c4c6d6f6b22097 --force (exit status 1)
   STDOUT -------->

   STDERR -------->

   SpecNotFound: /tmp/tmpop7mwsw6.yml is not a valid yaml file.


·· Failure creating environment for conda-py3.10-Cython-meson-python-numpy-pybind11-pytest-pythran
Traceback (most recent call last):
  File "/home/user/miniforge3/envs/scipy-dev/bin/asv", line 10, in <module>
    sys.exit(main())
  File "/home/user/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/asv/main.py", line 38, in main
    result = args.func(args)
  File "/home/user/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/asv/commands/__init__.py", line 49, in run_from_args
    return cls.run_from_conf_args(conf, args)
  File "/home/user/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/asv/commands/continuous.py", line 75, in run_from_conf_args
    return cls.run(
  File "/home/user/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/asv/commands/continuous.py", line 114, in run
    result = Run.run(
  File "/home/user/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/asv/commands/run.py", line 294, in run
    Setup.perform_setup(environments, parallel=parallel)
  File "/home/user/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/asv/commands/setup.py", line 89, in perform_setup
    list(map(_create, environments))
  File "/home/user/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/asv/commands/setup.py", line 21, in _create
    env.create()
  File "/home/user/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/asv/environment.py", line 704, in create
    self._setup()
  File "/home/user/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/asv/plugins/conda.py", line 174, in _setup
    self._run_conda(['env', 'create', '-f', env_file_name,
  File "/home/user/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/asv/plugins/conda.py", line 227, in _run_conda
    return util.check_output([conda] + args, env=env)
  File "/home/user/miniforge3/envs/scipy-dev/lib/python3.10/site-packages/asv/util.py", line 754, in check_output
    raise ProcessError(args, retcode, stdout, stderr)
asv.util.ProcessError: Command '/home/user/miniforge3/bin/conda env create -f /tmp/tmpop7mwsw6.yml -p /mnt/d/upstreams/scipy/benchmarks/env/dc4107ba04c9faab49c4c6d6f6b22097 --force' returned non-zero exit status 1
```

</details>

### PR DESCRIPTION
**Pull request by HaoZeke**
#### Reference issue

Closes https://github.com/scipy/scipy/issues/18849.
#### What does this implement/fix?

Switches to the latest ASV, this can be changed to a pinned release once `asv` has a new release. Also uses the faster `mamba` solver.
#### Additional information

N/A.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -77,8 +77,7 @@ jobs:
             pip install cython
             pip install numpy==1.23.5
             pip install -r doc_requirements.txt
-            # `asv` pin because of slowdowns reported in gh-15568
-            pip install mpmath gmpy2 "asv==0.4.2" pythran ninja meson click rich-click doit pydevtool pooch
+            pip install mpmath gmpy2 "git+https://github.com/airspeed-velocity/asv@d4c868e52943894688e2db6abbf1ba2ab179cb84" pythran ninja meson click rich-click doit pydevtool pooch
```
Comment (h-vetinari) [inline .circleci/config.yml:80]: Can you put asv on an extra line here please?

**Review by h-vetinari — COMMENTED** [review]
Thanks for the quick PR!

I picked this into my branch locally and it seems to work. 🥳 

Unfortunately, it seems that even the most current asv still suffers from a big enough performance regression that our job fails around ~70%:
```
[68.58%] ··· ...sform_mapping

context deadline exceeded

Try out a larger resource class or running tests in parallel to speed up job execution.  Upgrade your pricing plan to take advantage of longer max job runtimes.
```

**Comment by h-vetinari** [comment]
Also noticing the following while looking at the logs:
```
Couldn't load asv.plugins.mamba because
No module named 'mamba'
· Unknown environment type 'mamba'
```
Not sure if that's working as intended?

**Comment by h-vetinari** [comment]
Looking at the logs, I found a lot of failing benchmarks (unrelated to this PR, obviously); raised #18860

**Comment by h-vetinari** [comment]
Ah cool, great to hear that (lack of) `--quick` was the reason for the regression!

Now we're getting
```
[50.00%] ··· ...e.time_getrow ok
[50.00%] ··· ========= ======== ==========
              density   format            
             --------- -------- ----------
                0.05     csr     656±0μs  
                0.05     csc     12.1±0ms 
                0.05     lil     60.8±0ms 
                0.01     csr     627±0μs  
                0.01     csc     2.80±0ms 
                0.01     lil     17.4±0ms 
             ========= ======== ==========

real	15m24.635s
user	13m58.561s
sys	1m50.164s

Exited with code exit status 2
CircleCI received exit code 2
```

Seeing that this is right at the 50% mark, and how the benchmark suite _should_ fail (see #18868), could it be that you're hitting #18860 et al.? Or what does "exist status 2" mean?

File: .circleci/config.yml [hunk 2]
```diff
@@ -77,8 +77,9 @@ jobs:
             pip install cython
             pip install numpy==1.23.5
             pip install -r doc_requirements.txt
-            # `asv` pin because of slowdowns reported in gh-15568
-            pip install mpmath gmpy2 "asv==0.4.2" pythran ninja meson click rich-click doit pydevtool pooch
+            pip install mpmath gmpy2 \
+            "git+https://github.com/airspeed-velocity/asv@d4c868e52943894688e2db6abbf1ba2ab179cb84" \
+            pythran ninja meson click rich-click doit pydevtool pooch
```
Comment (h-vetinari) [inline .circleci/config.yml:82]: With "extra line", I meant
```suggestion
            pip install mpmath gmpy2 pythran ninja meson click rich-click doit pydevtool pooch
            pip install "git+https://github.com/airspeed-velocity/asv@d4c868e52943894688e2db6abbf1ba2ab179cb84"
```

**Comment by HaoZeke** [comment]
> Ah cool, great to hear that (lack of) `--quick` was the reason for the regression!
> 
> Now we're getting
> 
> ```
> [50.00%] ··· ...e.time_getrow ok
> [50.00%] ··· ========= ======== ==========
>               density   format            
>              --------- -------- ----------
>                 0.05     csr     656±0μs  
>                 0.05     csc     12.1±0ms 
>                 0.05     lil     60.8±0ms 
>                 0.01     csr     627±0μs  
>                 0.01     csc     2.80±0ms 
>                 0.01     lil     17.4±0ms 
>              ========= ======== ==========
> 
> real	15m24.635s
> user	13m58.561s
> sys	1m50.164s
> 
> Exited with code exit status 2
> CircleCI received exit code 2
> ```
> 
> Seeing that this is right at the 50% mark, and how the benchmark suite _should_ fail (see #18868), could it be that you're hitting #18860 et al.? Or what does "exist status 2" mean?



> Ah cool, great to hear that (lack of) `--quick` was the reason for the regression!
> 
> Now we're getting
> 
> ```
> [50.00%] ··· ...e.time_getrow ok
> [50.00%] ··· ========= ======== ==========
>               density   format            
>              --------- -------- ----------
>                 0.05     csr     656±0μs  
>                 0.05     csc     12.1±0ms 
>                 0.05     lil     60.8±0ms 
>                 0.01     csr     627±0μs  
>                 0.01     csc     2.80±0ms 
>                 0.01     lil     17.4±0ms 
>              ========= ======== ==========
> 
> real	15m24.635s
> user	13m58.561s
> sys	1m50.164s
> 
> Exited with code exit status 2
> CircleCI received exit code 2
> ```
> 
> Seeing that this is right at the 50% mark, and how the benchmark suite _should_ fail (see #18868), could it be that you're hitting #18860 et al.? Or what does "exist status 2" mean?

Exit status 2 is meant to signal a failed benchmark, but it should probably trigger a lot earlier.

**Comment by h-vetinari** [comment]
> What about something like Pytest where you can specify how many failure you want before raising?

I think the benchmarks should just fail if there are errors. We don't want green ci with any failing benchmarks (otherwise they bitrot hopelessly), and continuing to benchmark in the face of failures IMO also defeats the purpose.

This PR is great in that it basically solves #18868 en passant, but now we have to clean up all the failing benchmarks that have accumulated...

**Comment by tupui** [comment]
Right, agreed 👍

**Comment by h-vetinari** [comment]
@haozeke, thanks for your work on releasing the new asv 0.6! 🥳 

Given that this PR reveals breakage in the existing benchmarks, we're going to have to work a bit more on this. I have rebased this PR into https://github.com/scipy/scipy/pull/19052 for that purpose.

**Comment by HaoZeke** [comment]
Closing in favor of https://github.com/scipy/scipy/pull/19052

**Pull request closed without merging**